src/compiler/checker.ts

```diff
@@ -2014,6 +2017,34 @@ module ts {
             }
         }
 
+        // Push an entry on the type resolution stack. If an entry with the given target is not already on the stack,
+        // a new entry with that target and an associated result value of true is pushed on the stack, and the value
+        // true is returned. Otherwise, a circularity has occurred and the result values of the existing entry and
+        // all entries pushed after it are changed to false, and the value false is returned. The target object has
+        // no significance other than to provide a unique identity for a particular type resolution result.
+        function pushTypeResolution(target: Object): boolean {

```

Comment:
Also, it seems very easy to pass in the same object from two different callers by mistake, even when tracking different properties. I wish we could pass a stronger proxy for the property, rather than passing a different owner object in a way that is unrelated to which property we want to track.